### DISCUSSION TITLE
Is there anyway to sort the keys/values of the static map on device.

### DISCUSSION THREAD

**Discussion by Gibyeng** [NVIDIA/cuCollections#614] [Q&A] [answered]
I have n maps, each of them keeps some key-value pairs. I want to sort the values and the corresponding keys of them. This is an example:
Map[0]: {[0,2], [1,1], [2,4]}
Map[1]:{[0,4], [1,1], [2,5], [3,3]}
...
after sorting
arr[0]: {[2,4], [1,1], [0,2]}
arr[1]: { [2,5], [0,4], [3,3], [1,1]}
....
One solution is that for each of the maps, I use retrieve_all to save the key and value to a device_vector and then sort the vector one by one. However the number of maps is quite large. To make it more efficient, I am wondering how to sort all maps simultaneously?

**Comment by PointKernel** [comment, marked as answer]
The sorting issue is more relevant to [CCCL](https://github.com/NVIDIA/cccl).

One possible solution is to use `retrieve_all` to obtain all key-value pairs and then write a custom kernel where each block sorts a submap using [cub::DeviceRadixSort](https://nvidia.github.io/cccl/cub/api/structcub_1_1DeviceRadixSort.html#cub-deviceradixsort).

This can be further optimized by noting that each `retrieve_all` call launches a separate CUDA kernel. Instead, you can write a custom kernel to fuse all the "retrieve" operations into a single kernel.

**Reply by Gibyeng** [reply]
@PointKernel thanks for your advices.

**Reply by PointKernel** [reply]
@Gibyeng sorry, correction: [cub::BlockRadixSort](https://nvidia.github.io/cccl/cub/api/classcub_1_1BlockRadixSort.html#cub-blockradixsort) is the API you want for the block-wise sorting.
